buildSrc/src/main/java/quilt/internal/tasks/setup/DownloadMinecraftLibrariesTask.java

```diff
@@ -3,88 +3,143 @@
+            );
+        }));
 
-        getOutputs().dir(fileConstants.libraries);
-        outputsNeverUpToDate();
+        this.artifactsByUrl = this.getArtifactsByUrl();
     }
 
     @TaskAction
-    public void downloadMinecraftLibrariesTask() throws IOException {
-        Version file = Version.fromString(FileUtils.readFileToString(versionFile.get().getAsFile(), StandardCharsets.UTF_8));
+    public void downloadMinecraftLibrariesTask() {
+        this.getLogger().lifecycle(":downloading minecraft libraries");
 
-        getLogger().lifecycle(":downloading minecraft libraries");
+        this.getLibrariesDir().get().getAsFile().mkdirs();
 
-        if (!fileConstants.libraries.exists()) {
-            fileConstants.libraries.mkdirs();
-        }
+        final AtomicBoolean failed = new AtomicBoolean(false);
 
-        AtomicBoolean failed = new AtomicBoolean(false);
+        final Object lock = new Object();
 
-        Object lock = new Object();
+        final List<Library> librariesSrc = this.getVersion().get().getLibraries();
 
-        file.getLibraries().parallelStream().forEach(library -> {
-            Optional<DownloadableFile.PathDownload> artifact = library.getDownloads().getArtifact();
-            if (artifact.isEmpty()) {
-                return;
+        final Map<String, File> artifactsByUrl = librariesSrc.parallelStream().flatMap(library -> {
+            final Optional<DownloadableFile.PathDownload> artifactPath = library.getDownloads().getArtifact();
+            if (artifactPath.isEmpty()) {
+                return Stream.empty();
             }
 
+            final String url = artifactPath.get().getUrl();
+            final File artifact = this.artifactOf(url);
+
+            boolean thisFailed = false;
             try {
-                String url = artifact.get().getUrl();
-                startDownload()
-                        .src(url)
-                        .dest(getArtifactFile(fileConstants, url))
-                        .overwrite(false)
-                        .download();
+                this.startDownload()
+                    .src(url)
+                    .dest(artifact)
+                    .overwrite(false)
+                    .download();
             } catch (IOException e) {
-                failed.set(true);
+                thisFailed = true;
                 e.printStackTrace();
             } catch (NoSuchElementException e) {
                 new RuntimeException("Unable to find artifact for " + library.getName(), e).printStackTrace();
             }
 
             synchronized (lock) {
-                getProject().getDependencies().add("decompileClasspath", library.getName());
+                // TODO this is screwy.
+                //  Could we put these in a configuration in MappingsPlugin
+                //  and pass the configuration to an input of this task?
+                this.getProject().getDependencies().add("decompileClasspath", library.getName());
```

Comment:
So, basically -- you cannot, and should not, add dependencies at task execution time! You can expect all dependencies to be set up before then and need to set them up before then. You also should not pass around `Configuration`s as task inputs/properties used at task run time (config cache incompatible if nothing else) If you need this info at that time to do so -- perhaps a `ValueSource` could do it instead of a task? Or, if having it be a task is still easiest, have them saved in a file that's generated by a task ran when _`buildSrc`_ is built, so that they're available in the main project?